// curs/build/nodeTree/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeTree = void 0;
var IdBuilder = /** @class */ (function () {
    function IdBuilder() {
        this.id = 1;
    }
    IdBuilder.prototype.get = function () {
        return this.id++;
    };
    return IdBuilder;
}());
var idBuilder = new IdBuilder();
var NodeTree = /** @class */ (function () {
    function NodeTree(tag, attrs, parent) {
        this.id = idBuilder.get();
        this.parent = null;
        this.children = [];
        this.tag = tag;
        this.attrs = attrs;
        this.parent = parent ? parent : null;
        this.parent ? this.parent.add(this) : "";
    }
    NodeTree.prototype.add = function (nodeTree) {
        this.children.push(nodeTree);
    };
    NodeTree.prototype.remove = function (nodeTree) {
        for (var i = 0; i < this.children.length; i++) {
            var tree = this.children[i];
            if (tree.id === nodeTree.id) {
                this.children.splice(i, 1);
                return true;
            }
        }
        for (var _i = 0, _a = this.children; _i < _a.length; _i++) {
            var tree = _a[_i];
            if (tree.remove(nodeTree)) {
                return true;
            }
        }
    };
    return NodeTree;
}());
exports.NodeTree = NodeTree;
